### PR TITLE
feat: Migrate to TypeScript

### LINKED ISSUE #91: Migrate template to TypeScript
We've migrated https://github.com/netlify/zip-it-and-ship-it and https://github.com/netlify/functions to TypeScript and we've validated that the setup works well. We should migrate the template to use TypeScript, so that new projects are TypeScript by default.

### PR DESCRIPTION
**Pull request by netlify-team-account-1**
Closes #91 

This PR was modeled after https://github.com/netlify/zip-it-and-ship-it/pull/675.

### CODE REVIEW

**Comment by eduardoboucas** [comment]
@netlify-team-account-1 I see that this is adding a `package-lock.json` to the template, which doesn't currently exist. Was this intentional? It feels like we should let the lock file be generated on the first install, which will get the latest versions of all dependencies allowed by the ranges?

**Comment by netlify-team-account-1** [comment]
that was an accident, removed

File: {{cookiecutter.project_slug}}/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,70 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Basic Options */
+    "incremental": true,                   /* Enable incremental compilation */
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
+    "module": "commonjs",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
```
Comment (lukasholzer) [inline {{cookiecutter.project_slug}}/tsconfig.json:8]: We have `module` specified in the package.json https://github.com/netlify/node-template/blob/main/package.json#L6

So we should compile to ES Modules

 ```suggestion
    "module": "ESNext",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
```

File: {{cookiecutter.project_slug}}/tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    "sourceMap": false,                     /* Generates corresponding '.map' file. */
+    // "outFile": "./",                       /* Concatenate and emit output to single file. */
+    "outDir": "./dist",                        /* Redirect output structure to the directory. */
+    // "rootDir": "./",                       /* Specify the root directory of input files. Use to control the output directory structure with --outDir. */
+    // "composite": true,                     /* Enable project compilation */
+    // "tsBuildInfoFile": "./",               /* Specify file to store incremental compilation information */
+    "removeComments": false,                /* Do not emit comments to output. */
+    // "noEmit": true,                        /* Do not emit outputs. */
+    // "importHelpers": true,                 /* Import emit helpers from 'tslib'. */
+    "downlevelIteration": true,            /* Provide full support for iterables in 'for-of', spread, and destructuring when targeting 'ES5' or 'ES3'. */
```
Comment (lukasholzer) [inline {{cookiecutter.project_slug}}/tsconfig.json:24]: do we need downlevel iteration? I think we can safely support only es2017

```suggestion
```

File: {{cookiecutter.project_slug}}/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,70 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Basic Options */
+    "incremental": true,                   /* Enable incremental compilation */
+    "target": "es6",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (lukasholzer) [inline {{cookiecutter.project_slug}}/tsconfig.json:7]:
```suggestion
    "target": "ES2017",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```

As we are supporting `"^12.20.0 || ^14.14.0 || >=16.0.0"` we can use es2017

File: {{cookiecutter.project_slug}}/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
+
+    /* Additional Checks */
+    // "noUnusedLocals": true,                /* Report errors on unused locals. */
+    // "noUnusedParameters": true,            /* Report errors on unused parameters. */
+    // "noImplicitReturns": true,             /* Report error when not all code paths in function return a value. */
+    // "noFallthroughCasesInSwitch": true,    /* Report errors for fallthrough cases in switch statement. */
+
+    /* Module Resolution Options */
+    // "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```
Comment (lukasholzer) [inline {{cookiecutter.project_slug}}/tsconfig.json:44]: I think we need the moduleResultion here with esModuels

```suggestion
    "moduleResolution": "node",            /* Specify module resolution strategy: 'node' (Node.js) or 'classic' (TypeScript pre-1.6). */
```

**Review by lukasholzer — CHANGES_REQUESTED** [review]

File: {{cookiecutter.project_slug}}/tsconfig.json (hunk 1 above)
Comment (ehmicky) [reply {{cookiecutter.project_slug}}/tsconfig.json:8]: I am wondering whether `nodenext` should be used instead? Note: unfortunately, this is still in experimental mode.
When I tried using `"module": "esnext"` in ZISI when migrating to ES modules, this did not quite work, as I think that mode is meant for browsers, while `nodenext` is meant for pure ES modules in Node.js.

Comment (eduardoboucas) [reply {{cookiecutter.project_slug}}/tsconfig.json:8]: Doesn't `ESNext` mean features that are not on the spec yet?

File: {{cookiecutter.project_slug}}/tsconfig.json (hunk 3 above)
Comment (eduardoboucas) [reply {{cookiecutter.project_slug}}/tsconfig.json:7]: Sounds good!

**Comment by eduardoboucas** [comment]
Thanks for the review @lukasholzer @ehmicky! ❤️ 

@netlify-team-account-1 deferring back to you as this was your PR originally.

File: {{cookiecutter.project_slug}}/tsconfig.json (hunk 1 above)
Comment (ehmicky) [reply {{cookiecutter.project_slug}}/tsconfig.json:8]: In principle, yes. [The documentation mentions](https://www.typescriptlang.org/tsconfig#module) that `es2015` already outputs ES modules, but `es2020` adds dynamic `import` and `import.meta`, and `nodenext` adds pure ES modules support (`"type": "module" in `package.json`).

Comment (eduardoboucas) [reply {{cookiecutter.project_slug}}/tsconfig.json:8]: So it feels like `es2020` would be a sensible choice here?

Comment (netlify-team-account-1) [reply {{cookiecutter.project_slug}}/tsconfig.json:8]: yup, agreeing with `es2020`!

```suggestion
    "module": "es2020",                     /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
```

File: {{cookiecutter.project_slug}}/package.json [hunk 5]
```diff
@@ -3,12 +3,16 @@
   "private": true,
   "version": "{{cookiecutter.project_version}}",
   "description": "{{cookiecutter.project_description}}",
+  "type": "module",
+  "main": "./src/main.js",
+  "exports": "./src/main.js",
```
Comment (netlify-team-account-1) [inline {{cookiecutter.project_slug}}/package.json:8]:
```suggestion
  "main": "./dist/main.js",
  "exports": "./dist/main.js",
```

**Comment by netlify-team-account-1** [comment]
tested this out locally, seems to be working! :)

File: {{cookiecutter.project_slug}}/package.json [hunk 6]
```diff
@@ -4,32 +4,34 @@
...
-    "test:ci": "run-s test:ci:*",
+    "test:dev": "run-s build test:dev:*",
+    "test:ci": "run-s build test:ci:*",
     "test:dev:ava": "ava",
     "test:ci:ava": "nyc -r lcovonly -r text -r json ava"
   },
   "config": {
-    "eslint": "--ignore-path .gitignore --cache --format=codeframe --max-warnings=0 \"{src,scripts,.github}/**/*.{js,md,html}\" \"*.{js,md,html}\" \".*.{js,md,html}\"",
-    "prettier": "--ignore-path .gitignore --loglevel=warn \"{src,scripts,.github}/**/*.{js,md,yml,json,html}\" \"*.{js,yml,json,html}\" \".*.{js,yml,json,html}\" \"!**/package-lock.json\" \"!package-lock.json\""
+    "eslint": "--ignore-path .gitignore --cache --format=codeframe --max-warnings=0 \"{src,scripts,.github}/**/*.{js,ts,md,html}\" \"*.{js,ts,md,html}\" \".*.{js,ts,md,html}\"",
```
Comment (lukasholzer) [inline {{cookiecutter.project_slug}}/package.json:33]:
```suggestion
    "eslint": "--ignore-path .gitignore --cache --format=codeframe --max-warnings=0 \"{src,scripts,.github}/**/*.{js,cjs,mjs,ts,md,html}\" \"*.{js,cjs,mjs,ts,md,html}\" \".*.{js,cjs,mjs,ts,md,html}\"",
```

@ehmicky as we already have a common npm package for that is there a way to have that centralised?

**Review by lukasholzer — APPROVED** [review]
Awesome! Go TS 🥳

File: {{cookiecutter.project_slug}}/package.json (hunk 6 above)
Comment (ehmicky) [reply {{cookiecutter.project_slug}}/package.json:33]: Hum, good question. :thinking: 

The common package is for the ESLint configuration, which can be shared using the `extends` property. When it comes to CLI flags, we could only share by creating some binary in `@netlify/eslint-config-node` calling `eslint` with those flags, and then make our repository use that `eslint` wrapper.

**Review by ehmicky — APPROVED** [review]
:tada:

**Pull request merged**